File: 백준/Gold/2470. 두 용액/두 용액.py
import sys
sys.setrecursionlimit(100000)

input = sys.stdin.readline
output = sys.stdout.write

n = int(input())
lst = list(map(int,input().split()))
mn = 10000000001
com = [0,0]
lst.sort()
start = 0
end = n-1
while start < end:
  sum = lst[start] + lst[end]
  if mn > abs(sum):
    mn = abs(sum)
    com[0] = lst[start]
    com[1] = lst[end]
    if sum == 0:
      break
  if sum < 0:
    start += 1
  else:
    end -= 1

print(com[0],com[1])

File: 백준/Gold/2470. 두 용액/두 용액.cc
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    int n;
    vector<int> arr;
    cin >> n;
    for(int i = 0;i<n;i++){
        int k;
        cin >> k;
        arr.push_back(k);
    }
    sort(arr.begin(),arr.end());

    int start = 0;
    int end = n-1;
    long long sum = 10000000001;
    vector<int> save(2);
    while(start < end){
        long long result = arr[start] + arr[end];
        if(sum > abs(result)){
            sum = abs(result);
            save[0] = arr[start];
            save[1] = arr[end];
        }
        if(result < 0){
            start++;
        }
        else if(result == 0){
            save[0] = arr[start];
            save[1] = arr[end];
            break;
        }
        else{
            end--;
        }
    }
    sort(save.begin(),save.end());
    cout << save[0] << ' ' << save[1];
    return 0;
}